gateway-server/src/main/java/org/apache/knox/gateway/services/security/impl/RemoteAliasService.java

```diff
@@ -379,8 +378,44 @@ public void generateAliasForCluster(final String clusterName,
 
   @Override
   public char[] getGatewayIdentityPassphrase() throws AliasServiceException {
-    char[] passphrase = getPasswordFromAliasForGateway(
-        GATEWAY_IDENTITY_PASSPHRASE);
+    char[] passphrase = getPasswordFromAliasForGateway(config.getIdentityKeyPassphraseAlias());
+    if (passphrase == null) {
+      // Fall back to the keystore password if a key-specific password was not explicitly set.
+      passphrase = getGatewayIdentityKeystorePassword();
+    }
+    if (passphrase == null) {
+      // Use the master password if not password was found
+      passphrase = ms.getMasterSecret();
+    }
+    return passphrase;
+  }
+
+  @Override
+  public char[] getGatewayIdentityKeystorePassword() throws AliasServiceException {
+    char[] passphrase = getPasswordFromAliasForGateway(config.getIdentityKeystorePasswordAlias());
+    if (passphrase == null) {
+      passphrase = ms.getMasterSecret();
+    }
+    return passphrase;
+  }
+
+  @Override
+  public char[] getSigningKeyPassphrase() throws AliasServiceException {
```

Comment:
This is almost the same as `getGatewayIdentityPassphrase` except for the passphrase alias; consider creating a private method and invoke it using the proper parameters.

Reply:
I thought of this, but the fall back call `getSigningKeystorePassword` vs `getSigningKeystorePassword` would cause the need to provide some implemented interface.  I think that may be more work than the duplicated code.  However `getSigningKeystorePassword` and `getGatewayIdentityKeystorePassword` may all into this category for code reuse.